### DISCUSSION TITLE
Build extensions to ESM instead of CJS

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#18309] [Feature Requests]
### Summary

Right now extensions are built as CJS.

After #15190, I think it would be great if we build the extensions also to ESM instead of CJS.

Reference:

https://github.com/directus/directus/blob/f850bcbacdb6597a6110f06816c15f2a25b0a0f8/packages/extensions-sdk/src/cli/commands/build.ts#L619

### Motivation

Continue with #15190

### Detailed Design

TODO

### Must Have:
- Build extensions to ESM instead of CJS

### Drawbacks

TODO

### Alternatives

TODO

### Adoption Strategy

TODO

**Comment by rijkvanzanten** [comment]
@nickrum thoughts? 🙂

**Reply by nickrum** [reply]
Support for building extensions to ESM will be added in #18351

**Reply by abdonrd** [reply]
Amazing, thanks!

**Comment by abdonrd** [comment]
@rijkvanzanten related with this, when you create extensions with TypeScript:

`npm init directus-extension@latest` > `hook`/`operation`/... > `typescript`

The `tsconfig.json` is like this:

```json
{
  "compilerOptions": {
    "target": "ES2019",
    "lib": ["ES2019", "DOM"],
    "moduleResolution": "node",
    "strict": true,
    "noFallthroughCasesInSwitch": true,
    "esModuleInterop": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedParameters": true,
    "alwaysStrict": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictBindCallApply": true,
    "strictPropertyInitialization": true,
    "resolveJsonModule": false,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "allowSyntheticDefaultImports": true,
    "isolatedModules": true,
    "rootDir": "./src"
  },
  "include": ["./src/**/*.ts"]
}
```

Reference:
https://github.com/directus/directus/blob/main/packages/extensions-sdk/templates/common/typescript/config/tsconfig.json

**Reply by rijkvanzanten** [reply]
cc @nickrum

**Reply by paescuj** [reply]
cc @paescuj

**Reply by rijkvanzanten** [reply]
cc @rijkvanzanten huh what are we doing

**Reply by nickrum** [reply]
What's wrong with the `tsconfig.json` (aside from the fact that we should probably modernize it based on our new `tsconfig.json` that is used by core)?

**Reply by abdonrd** [reply]
Sorry, I should have been more explicit!

I expect to have:
`"moduleResolution": "NodeNext",`

Instead of:
`"moduleResolution": "node",`

If not, we need to import like (note the `dist`):
`import type { AbstractServiceOptions } from '@directus/api/dist/types';`

Because it's not following:
https://github.com/directus/directus/blob/bc66343e20dd5d2fdae2d5d53243ba2db15aa332/api/package.json#L53-L58

**Reply by nickrum** [reply]
Ah yeah, good one! I'll update this `tsconfig.json` to be in line with our internal one 👍

**Reply by paescuj** [reply]
> Ah yeah, good one! I'll update those `tsconfig.json`s to be in line with our internal one 👍

I'm not sure we want to enforce our project specific rules for extensions. Instead I think we should add https://github.com/tsconfig/bases as a devDep and ship it along with a file like:
```
{
  "extends": ["@tsconfig/node18/tsconfig.json", "@tsconfig/esm/tsconfig.json"],
  "compilerOptions": {
    "moduleResolution": "NodeNext"
  }
}
```

That way, users could easily adjust / extend it however they like, e.g. by updating it to:
```
{
  "extends": [
    "@tsconfig/node18/tsconfig.json",
    "@tsconfig/esm/tsconfig.json",
    "@tsconfig/strictest/tsconfig.json"
  ],
  "compilerOptions": {
    "moduleResolution": "NodeNext"
  }
}
```

For "app extensions" we could ship files similar to https://github.com/vuejs/create-vue/tree/main/template/tsconfig/base

(had some ideas around this, now we know why I "cc"-ed myself 😄)

**Edit:** We may also extend from our own "@directus/tsconfig" package by adding another reduced config.

**Reply by abdonrd** [reply]
> Ah yeah, good one! I'll update this `tsconfig.json` to be in line with our internal one 👍

Great, thanks @nickrum!

**Reply by abdonrd** [reply]
@paescuj what about the `moduleResolution` default? `node` or `NodeNext`?

**Reply by paescuj** [reply]
> @paescuj what about the `moduleResolution` default? `node` or `NodeNext`?

Sorry was supposed to be `NodeNext` 🤣 Updated my comment 👍

**Reply by nickrum** [reply]
Hmm 🤔 I kinda like the strict-by-default approach of our internal `tsconfig.json` to encourage users to use Typescript as it should be used (🧐) and the explicitness of seeing all options at once vs. having to search through some `@tsconfig/**/tsconfig.json` files in some repo.

**Comment by rijkvanzanten** [comment]
This should be resolved as https://github.com/directus/directus/pull/18351
